### PR TITLE
Add binary versioning

### PR DESCRIPTION
**Pull request by ezra-varady**
This addresses #239. It generates a version header that gets included in the binary and compared to the SQL API version. If they do not match it will refuse to interact with the index. Unfortunately the system catalog does not make extension information in C afaict, so we execute an actual query to get the SQL API version. The result is cached to minimize the overhead but this means that once the mismatch has been resolved the connection must be reset. I've tested it locally and it behaves correctly.

I also added a new class of tests to check that errors are generated during update tests. I don't love the setup, so if anyone has ideas please let me know. I added a new class of tests, `test-misc` that are intended for use in situations like this where we want to call a single test in a very specific instance. I couldn't reuse the parallel test because the expected output has to be different which was unfortunate. In addition to this I replaced various duplicated utils with symlinks

### CODE REVIEW

**Comment by ezra-varady** [comment]
it looks like benchmark is maybe checking out a tag without altering the version. The misc tests can't be used without bumping the version which is a bit annoying, I'll look into this tomorrow. Not sure what the deal with the missing gettext header is on osx either, it looks like it's installed?

File: scripts/test_updates.py [hunk 1]
```diff
@@ -10,6 +10,40 @@
...
 
+class Version:
+    def __init__(self, version: str):
+        self.version_numbers = [int(n) for n in version.split('.')]
+    def __lt__(self, other):
+        for i, v in enumerate(self.version_numbers):
+            if v < other.version_numbers[i]:
+                return True
+        return False
+    def __le__(self, other):
```
Comment (Ngalstyan4) [inline scripts/test_updates.py:21]: Can the rest be implemented in terms of `__lt__` to avoid repetition?

File: CMakeLists.txt [hunk 2]
```diff
@@ -203,6 +203,15 @@ add_custom_command(
 
 add_custom_target(phony_always_runs ALL DEPENDS ${CMAKE_BINARY_DIR}/${_script_file})
 
+add_custom_target(
+    generate_version_header
+    COMMAND ${CMAKE_SOURCE_DIR}/scripts/version_header.sh ${LANTERNDB_VERSION} ${CMAKE_BINARY_DIR}
```
Comment (Ngalstyan4) [inline CMakeLists.txt:208]: `cmake` has a built-in function for this kind of stuff (`configure_file`).
Is there a reason we are not using that?
(See `CONTROL_TEMPLATE` in this CmakeLists.txt for example usage)

File: src/hnsw/utils.c [hunk 3]
```diff
@@ -103,3 +109,65 @@ float4 *ToFloat4Array(ArrayType *arr)
...
+        text       *version_text;
+
+        if(SPI_connect() != SPI_OK_CONNECT) {
+            elog(ERROR, "could not connect to executor to check binary version");
+        }
+
+        query = "SELECT extversion FROM pg_extension WHERE extname = 'lantern'";
+
+        // Execute the query to figure out what version of lantern is in use in SQL
+        // todo: it would be nice to get this without actually executing a query
```
Comment (Ngalstyan4) [inline src/hnsw/utils.c:136]: it is fine to run the query, no worries!

File: src/hnsw/utils.h [hunk 4]
```diff
@@ -6,11 +6,15 @@
 #include "options.h"
 #include "usearch.h"
 
+extern bool versions_match;
```
Comment (Ngalstyan4) [inline src/hnsw/utils.h:9]: do these need to be exported? Isn't the `VersionMatch()` the only API used externally?

**Comment by ezra-varady** [comment]
I haven't been able to recreate the issue with the benchmark runner locally in the dev container. I'm not sure what's going on exactly since this error shouldn't be possible without an update and none is done

**Comment by ezra-varady** [comment]
It looks like maybe it was a caching issue? The test has started passing since I added and then removed the upterm job

**Review by Ngalstyan4 — APPROVED** [review]

**Review by var77 — APPROVED** [review]

**Pull request merged**